File: android/src/main/java/org/maplibre/reactnative/components/layers/MLRNLayerManager.java
package org.maplibre.reactnative.components.layers;

import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.ViewGroupManager;
import com.facebook.react.uimanager.annotations.ReactProp;

public abstract class MLRNLayerManager<T extends MLRNLayer<?>> extends ViewGroupManager<T> {

    protected abstract T createLayerViewInstance(ThemedReactContext reactContext);

    @Override
    protected T createViewInstance(ThemedReactContext reactContext) {
        return createLayerViewInstance(reactContext);
    }

    @ReactProp(name="id")
    public void setId(T layer, String id) {
        layer.setID(id);
    }

    @ReactProp(name="afterId")
    public void setAboveLayerID(T layer, String aboveLayerID) {
        layer.setAboveLayerID(aboveLayerID);
    }

    @ReactProp(name="beforeId")
    public void setBelowLayerID(T layer, String belowLayerID) {
        layer.setBelowLayerID(belowLayerID);
    }

    @ReactProp(name="layerIndex")
    public void setLayerIndex(T layer, int layerIndex){
        layer.setLayerIndex(layerIndex);
    }

    @ReactProp(name="minzoom")
    public void setMinZoomLevel(T layer, double minZoomLevel) {
        layer.setMinZoomLevel(minZoomLevel);
    }

    @ReactProp(name="maxzoom")
    public void setMaxZoomLevel(T layer, double maxZoomLevel) {
        layer.setMaxZoomLevel(maxZoomLevel);
    }

    @ReactProp(name="reactStyle")
    public void setReactStyle(T layer, ReadableMap style) {
        layer.setReactStyle(style);
    }
}

File: android/src/main/java/org/maplibre/reactnative/components/layers/MLRNLayerManager.kt
package org.maplibre.reactnative.components.layers

import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.ViewGroupManager
import com.facebook.react.uimanager.annotations.ReactProp

class MLRNLayerManager : ViewGroupManager<MLRNLayer>() {
    override fun getName(): String = REACT_CLASS

    override fun createViewInstance(reactContext: ThemedReactContext): MLRNLayer = MLRNLayer(reactContext)

    @ReactProp(name = "id")
    fun setId(
        layer: MLRNLayer?,
        id: String?,
    ) {
        layer!!.ID = id!!
    }

    @ReactProp(name = "layerType")
    fun setLayerType(
        layer: MLRNLayer?,
        layerType: String?,
    ) {
        layer!!.setLayerType(layerType)
    }

    @ReactProp(name = "source")
    fun setSource(
        layer: MLRNLayer?,
        sourceID: String?,
    ) {
        layer!!.setSourceID(sourceID)
    }

    @ReactProp(name = "sourceLayer")
    fun setSourceLayer(
        layer: MLRNLayer?,
        sourceLayerID: String?,
    ) {
        layer!!.setSourceLayerID(sourceLayerID)
    }

    @ReactProp(name = "afterId")
    fun setAfterId(
        layer: MLRNLayer?,
        aboveLayerID: String?,
    ) {
        layer!!.setAboveLayerID(aboveLayerID)
    }

    @ReactProp(name = "beforeId")
    fun setBeforeId(
        layer: MLRNLayer?,
        belowLayerID: String?,
    ) {
        layer!!.setBelowLayerID(belowLayerID)
    }

    @ReactProp(name = "layerIndex")
    fun setLayerIndex(
        layer: MLRNLayer?,
        layerIndex: Int,
    ) {
        layer!!.setLayerIndex(layerIndex)
    }

    @ReactProp(name = "minzoom")
    fun setMinzoom(
        layer: MLRNLayer?,
        minZoomLevel: Double,
    ) {
        layer!!.setMinZoomLevel(minZoomLevel)
    }

    @ReactProp(name = "maxzoom")
    fun setMaxzoom(
        layer: MLRNLayer?,
        maxZoomLevel: Double,
    ) {
        layer!!.setMaxZoomLevel(maxZoomLevel)
    }

    @ReactProp(name = "reactStyle")
    fun setReactStyle(
        layer: MLRNLayer?,
        style: ReadableMap?,
    ) {
        layer!!.setReactStyle(style)
    }

    @ReactProp(name = "filter")
    fun setFilter(
        layer: MLRNLayer?,
        filterList: ReadableArray?,
    ) {
        layer!!.setFilter(filterList)
    }

    companion object {
        const val REACT_CLASS: String = "MLRNLayer"
    }
}
